deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/Convolution3DLayer.java

```diff
@@ -0,0 +1,257 @@
+                    : "")
+                    + " " + layerId());
+        }
+
+
+        int miniBatch = input.size(0);
+        int inputChannels = isNCDHW ? input.size(1) : input.size(4);
+        int inD = isNCDHW ? input.size(2) : input.size(1);
+        int inH = isNCDHW ? input.size(3) : input.size(2);
+        int inW = isNCDHW ? input.size(4) : input.size(3);
+
+        int outWeightChannels = isNCDHW ? weights.size(0) : weights.size(4);
+        int inWeightChannels = isNCDHW ? weights.size(1) : weights.size(3);
+
+        if (inputChannels != inWeightChannels) {
+            String layerName = conf.getLayer().getLayerName();
+            if (layerName == null)
+                layerName = "(not named)";
+            throw new DL4JInvalidInputException("Cannot do forward pass in Convolution3D layer (layer name = "
+                    + layerName
+                    + ", layer index = " + index + "): number of input array channels does not match " +
+                    "CNN layer configuration"
+                    + " (data input channels = " + input.size(1)
+                    + ", [minibatch, inputChannels, depth, height, width]="
+                    + Arrays.toString(input.shape()) + "; expected" + " input channels = " + inWeightChannels + ") "
+                    + layerId());
+        }
+
+
+        int[] kernel = layerConfig.getKernelSize();
+        int[] dilation = layerConfig.getDilation();
+        int[] strides = layerConfig.getStride();
+
+        int[] pad;
+        int[] outSize;
+        if (mode == ConvolutionMode.Same) {
+            outSize = Convolution3DUtils.get3DOutputSize(
+                    input, kernel, strides, null, convolutionMode, dilation, isNCDHW);
+            int[] inSize = new int[]{inD, inH, inW};
+            pad = Convolution3DUtils.get3DSameModeTopLeftPadding(outSize,
+                    inSize, kernel, strides, dilation);
+        } else {
+            pad = layerConfig.getPadding();
+            outSize = Convolution3DUtils.get3DOutputSize(input, kernel, strides, pad, convolutionMode, dilation, isNCDHW);
+        }
+        int outD = outSize[0];
+        int outH = outSize[1];
+        int outW = outSize[2];
+
+        INDArray output = Nd4j.create(miniBatch * outWeightChannels * outH * outW * outD);
+        INDArray reshapedOutput;
+        if (isNCDHW)
+            reshapedOutput = output.reshape('c', miniBatch, outWeightChannels, outD, outH, outW);
+        else
+            reshapedOutput = output.reshape('c', miniBatch, outD, outH, outW, outWeightChannels);
+
+        int[] intArgs = new int[]{
+                kernel[0], kernel[1], kernel[2],
+                strides[0], strides[1], strides[2],
+                pad[0], pad[1], pad[2],
+                dilation[0], dilation[1], dilation[2],
+                mode == ConvolutionMode.Same ? 1 : 0,
+                isNCDHW ? 0 : 1
+        };
+
+        CustomOp op;
+        if (layerConfig.hasBias()) {
```

Comment:
Minor thing with this if statement: just have an INDArray[] inputArgs = ... in the if statement. Everything else in the builder there is exactly the same for both cases, and can be pulled out. That's a bit cleaner IMO.